File: frontend/feature/reports/src/main/kotlin/com/matrigluco/feature/reports/data/remote/ReportsApi.kt
package com.matrigluco.feature.reports.data.remote

import com.matrigluco.feature.reports.data.remote.dto.FileAssetDto
import com.matrigluco.feature.reports.data.remote.dto.ReportDto
import com.matrigluco.feature.reports.data.remote.dto.SaveReportRequestDto
import com.matrigluco.feature.reports.data.remote.dto.UpdateReportRequestDto
import okhttp3.MultipartBody
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Streaming

interface ReportsApi {

    @GET("api/v1/reports")
    suspend fun getReports(
        @Query("limit") limit: Int = 50
    ): List<ReportDto>

    @GET("api/v1/reports/{report_id}")
    suspend fun getReportDetail(
        @Path("report_id") reportId: String
    ): ReportDto

    @Multipart
    @POST("api/v1/files/upload")
    suspend fun uploadPrivateFile(
        @Part file: MultipartBody.Part
    ): FileAssetDto

    @POST("api/v1/reports")
    suspend fun saveReport(
        @Body payload: SaveReportRequestDto
    ): ReportDto

    @PATCH("api/v1/reports/{report_id}")
    suspend fun updateReport(
        @Path("report_id") reportId: String,
        @Body payload: UpdateReportRequestDto
    ): ReportDto

    @DELETE("api/v1/reports/{report_id}")
    suspend fun deleteReport(
        @Path("report_id") reportId: String
    ): Response<Unit>

    @Streaming
    @GET("api/v1/reports/{report_id}/pdf")
    suspend fun downloadReportPdf(
        @Path("report_id") reportId: String
    ): Response<ResponseBody>

    @Streaming
    @GET("api/v1/files/{file_id}/download")
    suspend fun downloadPrivateFile(
        @Path("file_id") fileId: String
    ): Response<ResponseBody>
}

File: web/src/features/reports/api/reports.api.ts
import { apiRequest } from "../../../services/http/client";
import {
  RawReportApiItem,
  UploadFileAssetResponse,
  UpdateReportExtractionPayload,
} from "../types/reports.types";

export const reportsApi = {
  /**
   * Retrieves list of medical reports for authenticated patient.
   */
  async getReports(
    params?: { limit?: number; page?: number },
    signal?: AbortSignal
  ): Promise<RawReportApiItem[]> {
    return apiRequest<RawReportApiItem[]>({
      url: "/reports",
      method: "GET",
      params,
      signal,
    });
  },

  /**
   * Retrieves single owned medical report detail.
   */
  async getReportById(
    reportId: string,
    signal?: AbortSignal
  ): Promise<RawReportApiItem> {
    return apiRequest<RawReportApiItem>({
      url: `/reports/${reportId}`,
      method: "GET",
      signal,
    });
  },

  /**
   * Uploads file to private healthcare storage with content signature validation.
   */
  async uploadPrivateFile(
    file: File,
    signal?: AbortSignal
  ): Promise<UploadFileAssetResponse> {
    const formData = new FormData();
    formData.append("file", file);

    return apiRequest<UploadFileAssetResponse>({
      url: "/files/upload",
      method: "POST",
      data: formData,
      headers: {
        "Content-Type": "multipart/form-data",
      },
      signal,
    });
  },

  /**
   * Creates a medical report record linking to the uploaded private file.
   */
  async saveReport(
    payload: {
      file_name?: string;
      file_url?: string;
      extracted_values?: Record<string, any>;
      prediction_result?: string;
      risk_level?: string;
    },
    signal?: AbortSignal
  ): Promise<RawReportApiItem> {
    return apiRequest<RawReportApiItem>({
      url: "/reports",
      method: "POST",
      data: payload,
      signal,
    });
  },

  /**
   * Updates an owned medical report's extracted biomarkers and review state.
   */
  async updateReport(
    reportId: string,
    payload: UpdateReportExtractionPayload,
    signal?: AbortSignal
  ): Promise<RawReportApiItem> {
    return apiRequest<RawReportApiItem>({
      url: `/reports/${reportId}`,
      method: "PATCH",
      data: payload,
      signal,
    });
  },

  /**
   * Deletes an owned medical report record.
   */
  async deleteReport(reportId: string, signal?: AbortSignal): Promise<void> {
    return apiRequest<void>({
      url: `/reports/${reportId}`,
      method: "DELETE",
      signal,
    });
  },

  /**
   * Downloads authorized private file stream as binary blob with no-store caching.
   */
  async downloadPrivateFileBlob(
    fileId: string,
    signal?: AbortSignal
  ): Promise<Blob> {
    return apiRequest<Blob>({
      url: `/files/${fileId}/download`,
      method: "GET",
      responseType: "blob",
      signal,
    });
  },

  /**
   * Downloads high-fidelity clinical PDF generated via Jinja2 HTML/CSS templates.
   */
  async downloadReportPdfBlob(
    reportId: string,
    signal?: AbortSignal
  ): Promise<Blob> {
    return apiRequest<Blob>({
      url: `/reports/${reportId}/pdf`,
      method: "GET",
      responseType: "blob",
      signal,
    });
  },
};
